corehq/apps/consumer_user/views.py

```diff
@@ -0,0 +1,189 @@
+from datetime import datetime
+
+from django.contrib import messages
+from django.contrib.auth import logout
+from django.contrib.auth.forms import PasswordChangeForm
+from django.contrib.auth.models import User
+from django.core.signing import BadSignature, SignatureExpired
+from django.http import HttpResponse, HttpResponseRedirect
+from django.shortcuts import render
+from django.urls import reverse
+from django.utils.decorators import method_decorator
+from django.utils.translation import ugettext as _
+from django.views.decorators.debug import sensitive_post_parameters
+
+from no_exceptions.exceptions import Http400
+from two_factor.forms import AuthenticationTokenForm, BackupTokenForm
+from two_factor.views import LoginView
+
+from corehq.apps.domain.decorators import two_factor_exempt
+from corehq.apps.users.models import CouchUser
+
+from .decorators import consumer_user_login_required
+from .forms.change_contact_details_form import ChangeContactDetailsForm
+from .forms.consumer_user_authentication_form import (
+    ConsumerUserAuthenticationForm,
+)
+from .forms.consumer_user_signup_form import ConsumerUserSignUpForm
+from .models import (
+    ConsumerUser,
+    ConsumerUserCaseRelationship,
+    ConsumerUserInvitation,
+)
+
+
+class ConsumerUserLoginView(LoginView):
+    form_list = (
+        ('auth', ConsumerUserAuthenticationForm),
+        ('token', AuthenticationTokenForm),
+        ('backup', BackupTokenForm),
+    )
+    invitation = None
+    signed_invitation_id = None
+    template_name = 'consumer_user/p_login.html'
+
+    @two_factor_exempt
+    @method_decorator(sensitive_post_parameters('password'))
+    def dispatch(self, request, *args, **kwargs):
+        if 'signed_invitation_id' in kwargs:
```

Comment:
I think this might be more sensible as a different view entirely - it's a separate URL with different args, and it's a one-time workflow (right?), versus the other usages, which are intended for regular use.

As-is, the logic to accept an invitation is shoehorned in the `ConsumerUserAuthenticationForm` (in the clean method), when it seems more like a separate concern.  Plus, most of the code in this view class is behind an `if invitation` clause, which doesn't apply most of the time.